### PR TITLE
macOS patches

### PR DESCRIPTION
**Pull request by nastys**
The macOS patches we've been working on on Discord.
This fork:branch is for reference — the actual patches will be uploaded as separate pull requests.

~~- [x] Building with GCC 11 (requires patching a couple of Qt libraries: [qcoreapplication.h](https://cdn.discordapp.com/attachments/272083901277667342/913925397333291018/qcoreapplication.h), [qtconcurrentthreadengine.h](https://cdn.discordapp.com/attachments/272083901277667342/913925397488496690/qtconcurrentthreadengine.h); and a couple of [3rdparty libraries](https://github.com/RPCS3/rpcs3/issues/5361#issuecomment-986430965); ffmpeg needs to be rebuilt)~~ I recommend using Homebrew clang/llvm instead, it's much more reliable
- [x] Building with Homebrew clang/llvm (``brew install llvm`` — it doesn't even need most of the patches in this fork, so you can build master instead; tested with and without the patched Qt libraries above)
- [x] Fast interpreter
- [x] SPU ASMJIT
- [x] FAudio
- [x] Keyboard input (arrows keys not working)
- [x] DS4 controller input (USB)
- [x] Threaded downloader (linker errors when building with GCC, but works fine with brew clang/llvm — currently patched not to use Thread on macOS in this fork, so it's recommended to build master instead)
- [x] Cubeb (``-DUSE_AUDIOUNIT=ON``, not working with GCC)
- [ ] SPU LLVM (unstable)
- [x] PPU LLVM (generates huge logs, for some reason)
- [x] Bluetooth controller input
- [x] CI
- [ ] DS3 controller input (USB; keeps flashing)
- [ ] Generic controller handler (USB)
- [ ] Discord integration (linker errors — untested with brew clang/llvm)
- [ ] Building with clang/Xcode SDK (some semaphore functions required by RPCN are not available, and Vulkan fails with "Unknown context dma 0x%x")

Graphics:
- [x] Vulkan via moltenVK (currently requires "select" to be renamed "sel":rmacs3/"selection":master due to a bug in SPIRV-Cross, and existing Apple patches for correct colour rendering)
- [x] Unsupported texture/colour surface format conversion (mostly done)
- [ ] VkLogicOp emulation
- [ ] On-Disk Shader Cache (must be disabled in Config -> Advanced)
- [ ] OpenGL (will probably never work — should probably be removed from macOS builds)

Games:
- [x] Project DIVA F
- [x] Project DIVA Dreamy Theater (missing projected shadows)
- [x] The Simpsons Game (minor graphical errors)
- [x] P****** 5 (~~only boots sometimes, maybe due to a race condition~~ seems to work reliably with brew clang/llvm)
- [x] Kingdom Hearts 2.5
- [x] Tekken Tag Tournament 2 (requires "Write Color Buffers" for some textures)

## Building
Build instructions are for x64. aarch64 may work with some changes once support for it is added to the emulator.
Install Homebrew.
Note that on M1, x64 Homebrew is required. Run the following command to install it:
```
arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
Afterwards, the following script can be used to install any x64 dependencies. Copy the following text to TextEdit or nano:
```
eval "$(/usr/local/Homebrew/bin/brew shellenv)"
arch -x86_64 brew "$@"
```
save it as ``~/brewx64.sh`` and run it like this to install the dependencies (one or two may be missing):
```
sh brewx64.sh install llvm sdl2 nasm molten-vk vulkan-headers qt@5 ninja glew cmake git
```
Env for building with Homebrew clang/llvm:
```
export CXX=clang++
export CC=clang
export Qt5_DIR="/usr/local/Cellar/qt@5/5.15.2_1/lib/cmake/Qt5"
export PATH="/usr/local/opt/llvm/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Library/Apple/usr/bin"
export LDFLAGS="-L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib"
export CPPFLAGS="-I/usr/local/opt/llvm/include -msse -msse2 -mcx16"
export CXXFLAGS="-I/usr/local/opt/llvm/include -msse -msse2 -mcx16"
export VULKAN_SDK="$HOME/VulkanSDK/1.2.198.1/macOS" # change the version if yours is different
export VK_ICD_FILENAMES="$VULKAN_SDK/share/vulkan/icd.d/MoltenVK_icd.json"
```
Clone the repository:
```
git clone --recursive -b rmacs3 https://github.com/nastys/rpcs3.git
```
Fix dependencies:
```
sed -i '' "s/if(APPLE)/if(CMAKE_C_COMPILER_ID MATCHES \"AppleClang\")/g" rpcs3/3rdparty/wolfssl/wolfssl/CMakeLists.txt
sed -i '' "s/extern const double NSAppKitVersionNumber;/const double NSAppKitVersionNumber = 1343;/g" rpcs3/3rdparty/hidapi/hidapi/mac/hid.c
```
Copy ``libc++abi.1.0.dylib`` from ``/usr/local/Cellar/llvm/13.0.0_2/lib`` to ``bin/rpcs3.app/Contents/Frameworks/libc++abi.1.dylib`` after building (if RPCS3 crashes).
Vulkan requires the [Vulkan SDK](https://vulkan.lunarg.com/) (if you get an error when you install it, just ignore it).
Finally:
```
mkdir rpcs3_build
cd rpcs3_build
cmake -DUSE_DISCORD_RPC=OFF -DUSE_VULKAN=ON -DUSE_ALSA=OFF -DUSE_PULSE=OFF -DUSE_AUDIOUNIT=ON -DCXXFLAGS="-msse -msse2 -mcx16" -G Ninja -DLLVM_CCACHE_BUILD=OFF -DLLVM_TARGETS_TO_BUILD="X86" -DLLVM_BUILD_RUNTIME=OFF -DLLVM_BUILD_TOOLS=OFF -DLLVM_INCLUDE_DOCS=OFF -DLLVM_INCLUDE_EXAMPLES=OFF -DLLVM_INCLUDE_TESTS=OFF -DLLVM_INCLUDE_TOOLS=OFF -DLLVM_INCLUDE_UTILS=OFF -DLLVM_USE_PERF=OFF -DLLVM_ENABLE_Z3_SOLVER=OFF -DCMAKE_CXX_STANDARD=20 ../rpcs3/
ninja
```
Use LLVM for PPU and ASMJIT for SPU until LLVM is fixed; disable On-Disk Shader Cache until fixed.
If you've previously attempted to use PPU LLVM (before it was fixed), you'll need to clear the old cache or it'll crash.
Also check the wiki for game-specific settings.

### CODE REVIEW

File: rpcs3/Emu/RSX/Program/GLSLCommon.cpp [hunk 1]
```diff
@@ -681,8 +681,14 @@ namespace glsl
 			"{\n"
 			"	vec4 low = cl * 12.92;\n"
 			"	vec4 high = 1.055 * pow(cl, vec4(1. / 2.4)) - 0.055;\n"
-			"	bvec4 select = lessThan(cl, vec4(0.0031308));\n"
-			"	return clamp(mix(high, low, select), 0., 1.);\n"
+      #ifdef __APPLE__
+            // temporary workaround for moltenVK
+			"	bvec4 sel = lessThan(cl, vec4(0.0031308));\n"
+			"	return clamp(mix(high, low, sel), 0., 1.);\n"
```
Comment (elad335) [inline rpcs3/Emu/RSX/Program/GLSLCommon.cpp:687]: I think all this shielding around this code is unjustified. It's simply a variable rename, it shouldn't cause any harm to other platforms.

Comment (kd-11) [reply rpcs3/Emu/RSX/Program/GLSLCommon.cpp:687]: Its a bug workaround. The variable can be omitted entirely by the time we're committing this, its just there because a bug report is open to track this externally I believe.

Comment (nastys) [reply rpcs3/Emu/RSX/Program/GLSLCommon.cpp:687]: It looks like a bug in SPIRV-Cross, but I haven't reported it yet because I wanted to fix it myself to be 100% sure, and also because I wanted to try more games.

Comment (kd-11) [reply rpcs3/Emu/RSX/Program/GLSLCommon.cpp:687]: It should be possible to reproduce using a trivial shader and send it to them. Just something a few lines long, I don't think it needs to be some complex rpcs3 shader to show the problem. They are usually quite fast at fixing these kinds of things.

**Comment by Vestrel** [comment]
Cubeb has native macOS api support. Idk why it tries to build with pulse/alsa.

**Comment by nastys** [comment]
> Cubeb has macOS native api support. Idk why it tries to build with pulse/alsa.

It does build with PulseAudio and ALSA disabled, but it fails to initialize.
```
E Cubeb: cubeb_init() failed: 0xffffffff
E SYS: Audio renderer Cubeb could not be initialized, using a Null renderer instead
```

**Comment by Vestrel** [comment]
Try enabling AudioUnit

File: rpcs3/Emu/RSX/Program/GLSLCommon.cpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/Program/GLSLCommon.cpp:687]: Just rename.
Upstream bugs are not a concern for our code.
It suffices if our code compiles and the upstream repo maintained are notified about the issue.

File: rpcs3/Emu/RSX/Program/GLSLCommon.cpp [hunk 2]
```diff
@@ -764,7 +770,7 @@ namespace glsl
 			OS <<
 
 #ifdef __APPLE__
-			"vec4 remap_vector(const in vec4 rgba, const in uint remap_bits)\n"
+			"vec4 remap_vector_m(const in vec4 rgba, const in uint remap_bits)\n"
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Program/GLSLCommon.cpp:773]: Why was this renamed?
Just wondering.

Comment (nastys) [reply rpcs3/Emu/RSX/Program/GLSLCommon.cpp:773]: It conflicted with the other remap_vector (both are required on macOS).

**Comment by adamtheman123** [comment]
IVE COME TO MAKE ANNOUNCEMNET I've come to make an announcement: MACOS's a bitch-ass motherfucker. He pissed on my fucking wife. That's right. He took his hedgehog fuckin' quilly dick out and he pissed on my FUCKING wife, and he said his dick was THIS BIG, and I said that's disgusting. So I'm making a callout post on my Twitter.com. Shadow the Hedgehog, you got a small dick. It's the size of this walnut except WAY smaller. And guess what? Here's what my dong looks like. That's right, baby. Tall points, no quills, no pillows, look at that, it looks like two balls and a bong. He fucked my wife, so guess what, I'm gonna fuck the earth. That's right, this is what you get! My SUPER LASER PISS! Except I'm not gonna piss on the earth. I'm gonna go higher. I'm pissing on the MOOOON! How do you like that, OBAMA? I PISSED ON THE MOON, YOU IDIOT! You have twenty-three hours before the piss DROPLETS hit the fucking earth, now get out of my fucking sight before I piss on you too!

**Comment by kode54** [comment]
If the .zip that `zip -r` is producing doesn't extract a working .app bundle, maybe try `ditto -c -k --sequesterRsrc --keepParent --zlibCompressionLevel 9` instead? It probably won't affect this package, but there's some Unicode normalization issue I remember there being with zip vs ditto.

**Comment by elliott44k** [comment]
> 

For the controller support, if it's OS 12 then I believe that this is a known issue currently.

Current short term fix is provided [https://github.com/libsdl-org/SDL/issues/4923#issuecomment-966722634](here).

**Comment by illusion0001** [comment]
https://github.com/libsdl-org/SDL/issues/4923#issuecomment-966722634 working url

**Comment by iMonZ** [comment]
Hey
Is it possible to create a brew formula for easier testing and to resolve the Notarization issue?

**Comment by iMonZ** [comment]
Question:
Would it be easier if RPCS3 would use SDL instead of cubeb?

**Comment by iMonZ** [comment]
@nastys why have you closed it?
It worked so great it could be merged.

**Comment by kode54** [comment]
@iMonZ If you compare the branches, the only difference between the main branch and nastys' macOS branch is that the main contains newer commits, and nastys' branch contains the rewrite of the CI scripts for macOS building. I'm not sure how much of it was unmerged before, but those CI changes still need to be applied to the main project. The branch still exists on their RPCS3 fork. It could probably use a rebase to get those newer changes.

It would be nice to know why the PR was closed instead of rebased, though.

**Comment by nastys** [comment]
Because these changes have already been merged.

**Comment by kode54** [comment]
And yet when I attempt to run a build produced from the latest master commit, it throws a "Verification failed" error for vec_alloc from simd.hpp.
